packages/grafana-runtime/src/utils/DataSourceWithBackend.ts

```diff
@@ -137,12 +139,19 @@ class DataSourceWithBackend<
           throw new Error(`Unknown Datasource: ${JSON.stringify(q.datasource)}`);
         }
 
-        datasource = ds.rawRef ?? getDataSourceRef(ds);
-        datasourceId = ds.id;
+        const dsRef = ds.rawRef ?? getDataSourceRef(ds);
+        const dsId = ds.id;
+        if (!eq(dsRef, datasource) || datasourceId !== dsId) {
```

Comment:
Interestenly enough, I only found that this condition is true (so `this` and `q.datasource` are different) when there is an `Expression` in the list of targets.

In the case of a `-- Mixed --` query, `this` is pointing to the right data source instance (as long as there is not an `Expression` query) on each iteration of the loop.

This issue would also be fixed if we modify the reason why `this` is pointing to the `ExpressionDatasource` rather than the actual data source but not sure if this condition is also true for other type of query.